src/docfx/build/toc/TableOfContentsMap.cs

```diff
@@ -164,25 +168,28 @@ public TocCandidate(int subDirectoryCount, int parentDirectoryCount, Document to
         /// <summary>
         /// Compare two toc candidate relative to target file.
         /// Return negative if x is closer than y, possitive if x is farer than y, 0 if x equals y.
-        /// 1. sub nearest
-        /// 2. parent nearest
-        /// 4. sub-name lexicographical nearest
+        /// 1. sub nearest(based on site path)
+        /// 2. parent nearest(based on site path)
+        /// 3. site path
+        /// 4. moniker intersection
+        /// 5. file path
         /// </summary>
-        private static int CompareTocCandidate(TocCandidate candidateX, TocCandidate candidateY)
+        private int CompareTocCandidate(TocCandidate candidateX, TocCandidate candidateY, Document file)
         {
-            var subDirCompareResult = candidateX.SubDirectoryCount - candidateY.SubDirectoryCount;
-            if (subDirCompareResult != 0)
+            var result = candidateX.SubDirectoryCount - candidateY.SubDirectoryCount;
+            if (result == 0)
+                result = candidateX.ParentDirectoryCount - candidateY.ParentDirectoryCount;
+            if (result == 0)
+                result = string.Compare(candidateX.Toc.SitePath, candidateY.Toc.SitePath, PathUtility.PathComparison);
+            if (result == 0 && _monikerProvider != null)
             {
-                return subDirCompareResult;
+                var fileMonikers = _monikerProvider.GetFileLevelMonikers(file).monikers;
+                result = _monikerProvider.GetFileLevelMonikers(candidateY.Toc).monikers.Intersect(fileMonikers, StringComparer.OrdinalIgnoreCase).Count() -
+                         _monikerProvider.GetFileLevelMonikers(candidateX.Toc).monikers.Intersect(fileMonikers, StringComparer.OrdinalIgnoreCase).Count();
             }
-
-            var parentDirCompareResult = candidateX.ParentDirectoryCount - candidateY.ParentDirectoryCount;
-            if (parentDirCompareResult != 0)
-            {
-                return parentDirCompareResult;
-            }
-
-            return candidateX.Toc.CompareTo(candidateY.Toc);
+            if (result == 0)
+                result = candidateX.Toc.FilePath.CompareTo(candidateY.Toc.FilePath);
```

Comment:
finally we still need compare `FilePath`?

Reply:
yes, when site/file relative path are both the same, we need to have a stable result